### ISSUE TITLE
KeyError while reading timeline event for gift feature

### ISSUE THREAD

**Issue by pleserwolf** [pytr-org/pytr#148]
**Description of the bug**
While downloading new files using pytr dl_docs, I now get a KeyError after all timeline events have been requested. It looks like response["id"] returns a UUID that is then not found in the timeline. The referenced UUID changes with every call of pytr dl_docs.

I have tried both running the command in my documents folder that is populated with a history of pytr documents as well as in a clean new folder. The only file that gets created in the new folder is an empty pytr_history.

Printing out response, it seems like the error is related to a transaction by the new "gift" feature.

```
{'id': 'REDACTED UUID', 'sections': [{'title': 'Du hast REDACTED\xa0€ geschenkt', 'data': {'icon': 'logos/contacts-M-Orange/v2', 'timestamp': 'REDACTED', 'status': 'executed'}, 'type': 'header'}, {'title': 'Übersicht', 'data': [{'title': 'Status', 'detail': {'text': 'Angenommen', 'functionalStyle': 'EXECUTED', 'type': 'status'}, 'style': 'plain'}, {'title': 'Asset', 'detail': {'text': 'REDACTED ETF NAME', 'type': 'text'}, 'style': 'plain'}, {'title': 'Lieferdatum', 'detail': {'text': '02 Dec', 'type': 'text'}, 'style': 'plain'}], 'type': 'table'}, {'title': 'Status', 'steps': [{'leading': {'avatar': {'status': 'completed', 'type': 'bullet'}, 'connection': {'order': 'first'}}, 'content': {'title': 'Geschenk geliefert', 'timestamp': 'REDACTED'}}, {'leading': {'avatar': {'status': 'completed', 'type': 'bullet'}, 'connection': {'order': 'last'}}, 'content': {'title': 'Geschenk angenommen', 'timestamp': 'REDACTED'}}], 'type': 'steps'}, {'title': 'Transaktion', 'data': [{'title': 'Geschenkbetrag', 'detail': {'text': 'REDACTED \xa0€', 'type': 'text'}, 'style': 'plain'}, {'title': 'Gebühr', 'detail': {
    'text': '1,00\xa0€', 'type': 'text'}, 'style': 'plain'}, {'title': 'Summe', 'detail': {'text': 'REDACTED\xa0€', 'type': 'text'}, 'style': 'highlighted'}], 'type': 'table'}, {'title': 'Empfänger', 'data': [{'title': 'REDACTED MAIL ADDRESS', 'detail': {'icon': 'logos/contacts-M-Orange/v2', 'style': 'highlighted', 'type': 'listItemAvatarDefault'}, 'style': 'highlighted'}], 'type': 'table'}, {'title': 'Mehr', 'data': [], 'type': 'table'}, {'title': 'Hilfe', 'data': [{'title': 'Problem melden', 'detail': {'icon': 'logos/timeline_communication/v2', 'action': {'type': 'customerSupportChat', 'payload': {'contextParams': {'chat_flow_key': 'NHC_0035_gifting_sent_a_gift', 'timelineEventId': 'REDACTED UUID', 'giftId': 'REDACTED UUID'}, 'contextCategory': 'NHC'}}, 'style': 'highlighted', 'type': 'listItemAvatarDefault'}, 'style': 'plain'}, {'title': 'Brauchst du Hilfe?', 'detail': {'icon': 'logos/timeline_document/v2', 'action': {'type': 'deeplink', 'payload': {'link': 'traderepublic://help-center/node/REDACTED'}}, 'style': 'highlighted', 'type': 'listItemAvatarDefault'}, 'style': 'plain'}], 'type': 'table'}]}

```

**To Reproduce**
I can reliably reproduce the error using my own TR account by running pytr dl_docs. It contains a gift transaction from yesterday.

**Error log**

```
15:40:03 Subscribing #2  timeline activity log
15:40:03 Received #2  timeline activity log
15:40:03 Received last relevant timeline activity log
15:40:03 All timeline details requested
Traceback (most recent call last):
  File "/Users/pleser/.local/bin/pytr", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/Users/pleser/.local/pipx/venvs/pytr/lib/python3.12/site-packages/pytr/main.py", line 266, in main
    asyncio.get_event_loop().run_until_complete(dl.dl_loop())
  File "/opt/homebrew/Cellar/python@3.12/3.12.7_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/pleser/.local/pipx/venvs/pytr/lib/python3.12/site-packages/pytr/dl.py", line 85, in dl_loop
    await self.tl.process_timelineDetail(response, self)
  File "/Users/pleser/.local/pipx/venvs/pytr/lib/python3.12/site-packages/pytr/timeline.py", line 130, in process_timelineDetail
    event = self.timeline_events[response["id"]]
            ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
KeyError: '20d2dd0d-7a1f-xxxx-xxxx-9a84e693dc0e'
```

**Environment**

- OS: macOS 15.1.1
- pytr version: 0.3.1

**Additional context**
I have pinned websockets to an older version as described in another issue.

**Comment by christoph-buente** [comment]
Same error here. Not sure if that has to do with the new gift feature they introduced or a general API upgrade.

**Comment by zonefuenf** [comment]
Could be! I updated the issue when I found that the download fails with that specific event, but it might be broader.

**Comment by xeo102** [comment]
The same case for me. The UUID for which the KeyError appears belongs to an outgoing gift.

**Comment by Jiter** [comment]
I can confirm that as well.
Also sent and received a gift. :)

**Comment by ProtectOurPlayground** [comment]
My wife and I have the same problem.
We also sent and received both a gift.
Before the "gift" thing, it worked.

**Comment by christoph-buente** [comment]
@Katzmann1983 @marzzzello I know it's annoying that things break when Trade Republic changes things in their API responses unexpectedly. But I would really appreciate it if you could look into this. Maybe there is a quick fix for it. (Even if it means it's ignoring the gift documents) but at least the download works again. Right now, the tool is essentially broken for all people who receive or send a gift. This is the gift of giving :)
Thank you! Really appreciate your work.

**Comment by zonefuenf** [comment]
I don't know any Python, but I tried putting together a small patch in PyCharm that just ignores unimplemented events, which works for me right now. 
[Ignore_unimplemented_events.patch](https://github.com/user-attachments/files/18081960/Ignore_unimplemented_events.patch)

**Comment by Austinpwrs** [comment]
@zonefuenf its not that the "event" is not supported, the issue is that the id of gift documents is different from the id of the gift in the timeline. 
In the doc you would have to match timelineEventId as you can see from @Jiter example payload with the id of the actual timeline event. Instead of matching the docs id with the id of the timeline event as the doc has its own id!

The next issue is that if you accept a gift another event is generated which again has to be matched - I didnt find any matching Ids here but I guess you can match by timestmap or something similar.

Then there is even a third new event - when you **win** a stock then you get an issue too.

![image](https://github.com/user-attachments/assets/c60de255-8c9c-437d-9137-f9465236db1e)


To summarize, 3 new events 
1. can be matched by timelineeventid
2. can be matched using timestamp maybe?
3. can not be matched I did not see a corresponding timelineevent to receiving a free stock

But this is all with just limited analysis

**Comment by HiidanX** [comment]
Has noone found a workaround? ptr dl_docs is not really usable right now.

**Comment by zonefuenf** [comment]
The patch I posted works as a workaround by ignoring the KeyErrors. It’s not a real solution, tho.

**Comment by HiidanX** [comment]
Is anyone working on a fix?

**Comment by JustinTrvz** [comment]
Here is my workaround to at least get the rest of the data. 
Try and catch the error to continue the workflow.

**pytr/dl.py**
_Line 80-93_
```python
if subscription.get("type", "") == "timelineTransactions":
    await self.tl.get_next_timeline_transactions(response)
elif subscription.get("type", "") == "timelineActivityLog":
    await self.tl.get_next_timeline_activity_log(response)
elif subscription.get("type", "") == "timelineDetailV2":
    try:
        await self.tl.process_timelineDetail(response, self)
    except Exception as e:
        self.log.error(f"Error processing timelineDetail: {e}")
        continue
else:
    self.log.warning(
        f"unmatched subscription of type '{subscription['type']}':\n{preview(response)}"
    )
```

**Comment by HiidanX** [comment]
Works, thank you very much.

**Comment by kacnje** [comment]
Do you think we can expect a solution soon? 
I can't help, I don't have enough knowledge.

**Comment by christoph-buente** [comment]
I've been using the [Software MoneyMoney](https://moneymoney-app.com/) for years and they introduced a feature for document downloads. This is how I resolved it for now, but I would be glad to get the functionality back.

**Comment by christoph-buente** [comment]
> Here is my workaround to at least get the rest of the data. Try and catch the error to continue the workflow.
> 
> **pytr/dl.py** _Line 80-93_
> 
> ```python
> if subscription.get("type", "") == "timelineTransactions":
>     await self.tl.get_next_timeline_transactions(response)
> elif subscription.get("type", "") == "timelineActivityLog":
>     await self.tl.get_next_timeline_activity_log(response)
> elif subscription.get("type", "") == "timelineDetailV2":
>     try:
>         await self.tl.process_timelineDetail(response, self)
>     except Exception as e:
>         self.log.error(f"Error processing timelineDetail: {e}")
>         continue
> else:
>     self.log.warning(
>         f"unmatched subscription of type '{subscription['type']}':\n{preview(response)}"
>     )
> ```

@JustinTrvz Would you mind sending this as a pull request?

**Comment by JustinTrvz** [comment]
@christoph-buente I would not because this is a **workaround** not a solution.

**Comment by ovizii** [comment]
Hm, I just tried, and these errors are gone, but I am seeing a new error now. Is this somehow related or a different issue by any chance?

```
10:01:27  991/5987: INM KFL SB SELIMB -- None - 2024-08-16T16:04:21
Traceback (most recent call last):
  File "/app/bin/pytr", line 8, in <module>
    sys.exit(main())
  File "/app/lib/python3.9/site-packages/pytr/main.py", line 266, in main
    asyncio.get_event_loop().run_until_complete(dl.dl_loop())
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/app/lib/python3.9/site-packages/pytr/dl.py", line 85, in dl_loop
    await self.tl.process_timelineDetail(response, self)
  File "/app/lib/python3.9/site-packages/pytr/timeline.py", line 130, in process_timelineDetail
    event = self.timeline_events[response["id"]]
KeyError: '46b04d75-5623-4967-ac6a-2680b2cf3c46'
```

**Comment by JustinTrvz** [comment]
@ovizii Please checkout the latest version of this repository by using `git pull` and verify if the error still occurs.

**Comment by ovizii** [comment]
> [@ovizii](https://github.com/ovizii) Please checkout the latest version of this repository by using `git pull` and verify if the error still occurs.

Sorry, I realized it's my fault as I relied on `pip install pytr` and I guess it doesn't reflect there yet.

**Comment by NiklasRosenstein** [comment]
A new release hasn't been cut yet, sorry for the confusion. I should've mentioned that!
